accounting/accounting_cli_functions.py

```diff
@@ -248,6 +248,196 @@ def view_jobs_before_end_time(conn, time_before, output_file):
     return job_records
 
 
+def add_bank(conn, bank, shares, parent_bank=""):
+    parent_bank_shares = 0
+    total_sub_bank_shares = 0
+
+    # if the parent bank is not "", that means the account
+    # trying to be added wants to be placed under a parent bank
+    if parent_bank != "":
+        try:
+            select_stmt = "SELECT shares FROM bank_table where bank=?"
+            dataframe = pd.read_sql_query(select_stmt, conn, params=(parent_bank,))
+            # if length of dataframe is 0, that means the parent bank wasn't found
+            if len(dataframe.index) == 0:
+                print("Parent account not found in bank table")
+                sys.exit(-1)
+            # fetch parent account from account
+            for index, row in dataframe.iterrows():
+                parent_bank_shares = row["shares"]
+            if int(shares) > int(parent_bank_shares):
+                print("Shares must not be greater than parent account's shares")
+                sys.exit(-1)
+
+            select_stmt = "SELECT shares FROM bank_table where parent_bank=?"
+            dataframe = pd.read_sql_query(select_stmt, conn, params=(parent_bank,))
+            # if length of dataframe is 0, that means there are no sub banks for
+            # this parent bank
+            if len(dataframe.index) == 0:
+                pass
+            # add up all of the shares allocated to this parent's sub banks
+            else:
+                for index, row in dataframe.iterrows():
+                    total_sub_bank_shares += row["shares"]
+            # if the total amount of shares in all the sub banks is
+            # equal to the parent bank's shares, it has already reached
+            # its max allocation, and thus the bank won't be added
+            if int(total_sub_bank_shares) >= int(parent_bank_shares):
+                print("Total shares for parent account are already maxed")
+                sys.exit(-1)
+            # if the current total amount of shares in all the sub banks
+            # PLUS the number of shares trying to be added is greater than the
+            # parent's banks shares, the parent account WILL exceed its max
+            # allocation and thus the account won't be added
+            if int(total_sub_bank_shares) + int(shares) > int(parent_bank_shares):
+                print("Total shares will exceed parent account's allocation")
+                sys.exit(-1)
```

Comment:
I will admit that my familiarity with how we plan on implementing shares is very little, but it appears that the constraints enforced here (i.e., `parent shares >= sum(child shares)`) is not enforced by LLNL's existing banks: https://computing.llnl.gov/tutorials/moab/#Banks

It looks like the existing bank algorithms just apply a per-level normalization.  Do we instead plan to enforce a total share limit?  Is this a constraint that Ryan or Greg asked for?  If so, I think we may want to enable fractional shares to enable finer grained splitting (i.e., splitting 10 shares across 100 users).

Reply:
Thanks for this explanation and the reference - the `parent shares >= sum(child shares)` constraint should not be included here. My plan was just to emulate the current system. I'll remove that check now.

Reply:
> the parent shares >= sum(child shares) constraint should not be included here.

As far as I can tell, none of the share constraints exist in LLNL's existing banks.  The referenced example has a bank (`bank1`) with 3 shares that has children with 16, 48, and 32 shares each.

That being said, I ran `sshare` on a production cluster, and it looks like by convention that most every bank/account follows the constraints that you have guaranteed.  The only exceptions are the "expired" bank (which has all of the old accounts in it) and this one:
```
❯ sshare -A lprpp,lprpp-1,lprpp-2
             Account       User  RawShares  NormShares    RawUsage  EffectvUsage  FairShare
-------------------- ---------- ---------- ----------- ----------- ------------- ----------
lprpp                                 1945    0.026231  1375628740      0.133639
 lprpp-1                              1588    0.500000  1375628694      1.000000
 lprpp-2                              1588    0.500000          45      0.000000
```

In summary, I think the code either include all of the constraints that you initially had or none of them.  In the interest of getting this PR merged as soon as possible, I'm fine with having them all in here, and we can discuss this further in an issue.  Given that, do you mind adding the `parent shares >= sum(child shares)` constraint back in?

Reply:
This seems like a good plan. I don't mind at all - just pushed an update that adds that constraint back in.